### PR TITLE
Generators: FP8 emission + expert unfusing + download tool

### LINKED ISSUE #238: Test-model generators: FP8 emission + expert weight unfusing (fixes silent expert-skip bug in converter)
## TL;DR

The test-model generators (`make_glm_oracle.py`, `make_glm_bench_model.py`, `glm_fp8_emit.py`) now:
1. Emit weights as **FP8 e4m3 + 128×128 block scale_inv** in the exact layout of the real GLM-5.2-FP8 checkpoint, so the converter's FP8→int4 dequant path can be exercised locally without the 379 GB download.
2. **Unfuse** HF's 3-D `gate_up_proj` expert weights into the per-expert 2-D layout (`gate_proj`/`up_proj`/`down_proj`) that the real checkpoint, the converter, and the C engine all expect.

Two bugs were found and fixed in the process.

Branch: `windows-dev` (commits `4788685`, `91b1035`, `d9b6a2d`)

---

## Why

The converter (`convert_fp8_to_int4.py`) has a dequant path for FP8 e4m3 + 128×128 block scales. Before this change, the only way to exercise that path was to download the 379 GB GLM-5.2-FP8 checkpoint. There was no local fixture that could validate the FP8→int4 pipeline, meaning bugs in the dequant path would only surface during the multi-hour production conversion.

### The FP8 emission gap

`glm_fp8_emit.py` is a new shared helper that:
- Block-quantizes 2-D tensors to FP8 e4m3 with 128×128 block scales (FBGEMM/TransformerEngine convention: `scale = amax(block)/448`, stored as a multiplier)
- Keeps norms, router, 1-D biases, and 3-D tensors at f32 (matching the converter's `classify()` + `ndim != 2` guard)
- Provides `fp8_block_dequantize` (exact inverse) for the oracle's reference round-trip

Both generators gained a `--fp8` flag. The oracle additionally round-trips weights through FP8 *before* computing `ref_glm.json`, so the reference reflects exactly the FP8 model the converter ingests.

### Bug 1: Fused expert weights (found during validation)

HF's `GlmMoeDsaForCausalLM` fuses `gate_proj` + `up_proj` into a single 3-D `gate_up_proj` tensor `[E, 2*M, I]` for compute efficiency. But the **real GLM-5.2-FP8 checkpoint stores experts unfused** as per-expert 2-D tensors:

```
model.layers.10.mlp.experts.0.gate_proj.weight       (2048, 6144)   FP8 e4m3
model.layers.10.mlp.experts.0.gate_proj.weight_scale_inv  (16, 48)   F32
model.layers.10.mlp.experts.0.up_proj.weight          (2048, 6144)   FP8 e4m3
model.layers.10.mlp.experts.0.up_proj.weight_scale_inv   (16, 48)   F32
model.layers.10.mlp.experts.0.down_proj.weight        (6144, 2048)   FP8 e4m3
model.layers.10.mlp.experts.0.down_proj.weight_scale_inv (48, 16)    F32
```

The converter's `classify()` matches `".mlp.experts." in name`, but its `ndim != 2` guard silently skipped the 3-D fused tensors. The C engine then crashed with `missing tensor: model.layers.3.mlp.experts.5.gate_proj.weight` because it expects per-expert 2-D tensors.

**Fix**: `unfuse_experts()` in `glm_fp8_emit.py` splits `gate_up_proj [E, 2M, I]` → per-expert `gate_proj [M, I]` + `up_proj [M, I]`, and `down_proj [E, I, M]` → per-expert `down_proj [M_out, I]`. Called after reference generation (the model needs fused weights for `forward()`/`generate()`) but before saving, in both generators, for both FP8 and bf16 paths.

### Bug 2: Oracle FP8 round-trip guard

`make_glm_oracle.py`'s FP8 round-trip loop used `p.dim() < 2` to skip non-quantizable tensors. But 3-D fused expert tensors have `dim() == 3`, which passed the guard and crashed `fp8_block_quantize` (which expects exactly 2-D). Fixed to `p.dim() != 2`, matching the converter's guard.

## What changed

| File | Change |
|------|--------|
| `c/tools/glm_fp8_emit.py` | New: `fp8_block_quantize`, `fp8_block_dequantize`, `keep_f32`, `state_dict_to_fp8`, `save_fp8_safetensors`, `unfuse_experts` |
| `c/tools/make_glm_oracle.py` | `--fp8` flag; round-trips weights through FP8 before computing reference; unfuses experts before saving; fixed dim guard |
| `c/tools/make_glm_bench_model.py` | `--fp8` flag; unfuses experts before saving; writes config.json manually in FP8 path (bypasses `save_pretrained`) |

## Impact on the program

- **Enables local testing of the full conversion chain**: generator `--fp8` → converter `--indir` → engine load, all without a network download.
- **Fixes a silent data-loss bug**: without unfusing, the converter would skip all expert weights and the engine would crash. This would have manifested as a broken production conversion once the FP8 download completed.
- **Backward compatible**: without `--fp8`, both generators behave exactly as before (bf16 output, `save_pretrained` path).
- **Validated end-to-end**: 570 FP8 tensors + 629 scale_inv sidecars → converter produces per-row int4 (fmt=2) and grouped int4 (fmt=4) → engine loads both clean, fmt=4 auto-detected.

## Verification

```
generator --fp8 → 570 e4m3 tensors + 629 scale_inv (was 90 when fused)
converter --group-size 0   → per-row int4 (fmt=2), engine loads clean, 0 missing-tensor errors
converter --group-size 128 → grouped int4 (fmt=4), 8-16x more scales per tensor, engine loads clean
dequant error: grouped 1.14-1.22x lower than per-row vs FP8 source
```

### LINKED ISSUE #239: FP8 download tool: dual-source (ModelScope + HF), parallel, resume — for the 379 GB GLM-5.2-FP8 checkpoint
## TL;DR

New `download_fp8.py` tool for downloading the GLM-5.2-FP8 checkpoint (379 GB, 141 shards) with **dual-source fallback** (ModelScope primary, HuggingFace fallback), parallel shard download, stall recovery, and clean resume. Designed for the bandwidth-constrained, throttle-prone reality of downloading a 379 GB model.

Branch: `windows-dev` (commit `91b1035`)

---

## Why

The production int4 conversion requires the FP8 source checkpoint. At 379 GB across 141 shards, the download is the **multi-day bottleneck** of the entire pipeline. Two problems with the default HuggingFace download path:

1. **HuggingFace per-stream throttling**: HF rate-limits individual streams after ~50 GB, dropping throughput to a crawl. There's no documented workaround for unauthenticated bulk downloads.
2. **No clean resume**: `hf_hub_download` resumes by file, but a mid-shard interruption leaves a partial file that may or may not be revalidated correctly, and there's no visibility into which shards are complete.

ModelScope mirrors the same checkpoint (`ZhipuAI/GLM-5.2-FP8`) on a CDN that doesn't apply the same per-stream throttling, giving 3-5x higher sustained throughput in practice.

## What it does

- **Dual-source**: tries ModelScope first (`ZhipuAI/GLM-5.2-FP8`); if the API is unreachable, falls back to HuggingFace (`zai-org/GLM-5.2-FP8`). Force a source with `--source ms` or `--source hf`.
- **Parallel download**: N worker threads (default 3) pull shards concurrently from a shared work queue.
- **Resume by size**: skips any shard whose local file size matches the expected size. Partial shards are re-downloaded from scratch (shards are 5 GB each, so re-downloading one is cheap).
- **Stall recovery**: 3 retry attempts per shard with exponential backoff; curl fallback (`-C -` for byte-range resume) if the Python download path fails.
- **Progress display**: per-shard completion with running percentage, elapsed time, and ETA.

## Usage

```bash
# Auto (ModelScope first, HF fallback), 3 parallel
python download_fp8.py

# 4 parallel, force ModelScope
python download_fp8.py --parallel 4 --source ms

# Force HuggingFace
python download_fp8.py --source hf
```

Re-running resumes — completed shards are skipped by size, only missing/partial ones are re-downloaded.

## Impact on the program

- **Unblocks the production conversion**: this is the prerequisite for the grouped-quantization quality A/B (issue #225) and the mixed-precision conversion (#237).
- **No engine changes**: the tool is a standalone download helper. It doesn't touch the C engine or the converter.
- **Standalone**: no dependencies beyond `huggingface_hub`, `modelscope`, and `requests` (all already in the project's Python environment).

## Status

Download is in progress (79/141 shards complete as of this writing). Re-running the tool resumes from where it left off.

### PR DESCRIPTION
**Pull request by woolcoxm**
## Summary

Three changes to the test-model generators:
1. **FP8 e4m3 emission** — generators can now write weights as FP8 + 128×128 block scales in the exact layout of the real GLM-5.2-FP8 checkpoint, so the converter's FP8→int4 dequant path can be tested locally without the 379 GB download.
2. **Expert weight unfusing** — splits HF's fused 3-D `gate_up_proj` into the per-expert 2-D layout the real checkpoint, converter, and C engine all expect. Fixes a silent data-loss bug.
3. **FP8 download tool** — `download_fp8.py` with ModelScope+HF dual-source, parallel shards, resume.

Closes #238, closes #239.

## Problem 1: No local FP8 fixture

The converter's FP8→int4 dequant path could only be exercised by downloading the 379 GB checkpoint. There was no way to test it locally, meaning dequant bugs would only surface during the multi-hour production conversion.

**Fix**: New `glm_fp8_emit.py` helper with `fp8_block_quantize` / `fp8_block_dequantize` / `save_fp8_safetensors`. Both generators gained `--fp8` flag. The oracle round-trips weights through FP8 before computing the reference.

## Problem 2: Fused expert weights (bug)

HF's `GlmMoeDsaForCausalLM` fuses `gate_proj` + `up_proj` into a 3-D `gate_up_proj [E, 2M, I]`. But the real checkpoint stores experts **unfused** as per-expert 2-D tensors:

```
model.layers.10.mlp.experts.0.gate_proj.weight       (2048, 6144)  FP8 e4m3
model.layers.10.mlp.experts.0.gate_proj.weight_scale_inv  (16, 48)  F32
model.layers.10.mlp.experts.0.up_proj.weight          (2048, 6144)  FP8 e4m3
...
```

The converter's `ndim != 2` guard **silently skipped** the 3-D fused tensors. The engine then crashed with `missing tensor: model.layers.3.mlp.experts.5.gate_proj.weight`.

**Fix**: `unfuse_experts()` in `glm_fp8_emit.py` splits `gate_up_proj [E, 2M, I]` → per-expert `gate_proj [M, I]` + `up_proj [M, I]`, and `down_proj [E, I, M]` → per-expert `down_proj`. Called after reference generation but before saving, in both generators.

Also fixed: `make_glm_oracle.py` FP8 round-trip guard `p.dim() < 2` → `p.dim() != 2` (3-D fused experts crashed `fp8_block_quantize`).

## Problem 3: Download tool

The 379 GB FP8 download is the pipeline bottleneck. HuggingFace per-stream throttling drops throughput after ~50 GB. `download_fp8.py` tries ModelScope first (no throttling), falls back to HF, with parallel shards + resume.

## Changes

| File | Change |
|------|--------|
| `c/tools/glm_fp8_emit.py` | NEW: FP8 block quantize/dequantize, `unfuse_experts`, `save_fp8_safetensors` |
| `c/tools/make_glm_oracle.py` | `--fp8` flag, FP8 round-trip, unfuse, dim guard fix |
| `c/tools/make_glm_bench_model.py` | `--fp8` flag, unfuse, manual config.json in FP8 path |
| `c/download_fp8.py` | NEW: dual-source parallel download with resume |

## Validation

```
generator --fp8 → 570 e4m3 tensors + 629 scale_inv (was 90 when fused)
converter --group-size 0   → per-row int4 (fmt=2), engine loads clean
converter --group-size 128 → grouped int4 (fmt=4), engine loads clean
dequant error: grouped 1.14-1.22x lower than per-row
```

## Test plan

- [x] Generator `--fp8` produces valid FP8 checkpoint with per-expert 2-D tensors
- [x] Converter ingests FP8 fixture, produces int4 output
- [x] Engine loads converted model with 0 missing-tensor errors
- [x] FP8 round-trip: dequant matches quant (maxdiff 0.0)
- [x] Download tool resumes (79/141 → in progress)

### CODE REVIEW

**Comment by JustVugg** [comment]
Closing — like #243, **this is already on `dev`**, arrived via your own earlier commits.

I checked each file rather than guessing:

- `c/download_fp8.py` — **identical** to `dev` (came in with 69f65d5)
- `c/tools/glm_fp8_emit.py` — **identical** to `dev` (71c262c and 2c6946c)
- `c/tools/make_glm_bench_model.py` — **identical** to `dev`
- `c/tools/make_glm_oracle.py` — the only file that differs, and it differs because **`dev` is ahead of this branch**: it has #318's `glm_tiny/` mkdir and #322's transformers version gate on top of your `--fp8` work. That's why this PR now shows as conflicting: not because it's wrong, because it's behind itself.

Zero unique content left. The `--fp8` path, the ref-after-round-trip design (computing `ref_glm.json` **after** the FP8 round-trip so the reference matches what the converter ingests — that's a careful piece of thinking, and exactly the trap #281 fell into from the other direction) and `unfuse_experts` are all live on `dev`.

Also closing #238 (FP8 emission + unfusing) and #239 (dual-source download tool) — I verified `download_fp8.py` has ModelScope + HF, resume, and parallel workers, so both are delivered.

Nothing lost here except the bookkeeping.
